### PR TITLE
JNLP VersionId and VersionString implementation

### PR DESCRIPTION
**Pull request by AndreasEhret**
- set existing JNLP Version implementation to deprecated
- provide new VersionId and VersionString implementations that cover JSR 56 Appendix A / Chapter 3.4 requirements on versions, including compound version handling
- new implementations addressing the TODOs in the current ITW Version class
- provided exhaustive tests for documentation and JSR specification compliance verification

### CODE REVIEW

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/JNLPVersionSpecifications.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    //
+    // This specification is implemented by the following regular expressions:
+    //
+    String REGEXP_SPACE = "[\\s]";
+    String REGEXP_SIMPLE_RANGE = "(" + REGEXP_VERSION_ID + ")";
+    String REGEXP_VERSION_RANGE = "(" + REGEXP_SIMPLE_RANGE + "(&" + REGEXP_SIMPLE_RANGE + ")*)";
+    String REGEXP_VERSION_STRING = "(" + REGEXP_VERSION_RANGE + "((" + REGEXP_SPACE + ")" + REGEXP_VERSION_RANGE + ")*)";
+
+    // Separators
+    enum Separator {
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/JNLPVersionSpecifications.java]: agree. Make them top level and rename to VersionSeparator and VersionModifier.

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    private final String versionId;
+    private final VersionId compoundVersionId;
+
+    private VersionId(String versionId, VersionId compoundVersionId) {
+        this.versionId = versionId;
+        this.compoundVersionId = compoundVersionId;
+    }
+
+    boolean hasPrefixMatchModifier() {
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: right now package private as there are tests for it. Will make them public, there is nothing wrong with it asking a VersionId if it has a prefix or greaterThanOrEqualMatch modifier.

[the remark this answers is not in the document]
Comment (hendrikebbers) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: yes, would prefer to make it public in that case

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+            return normalizedTuple;
+        }
+        else {
+            return tuple;
+        }
+    }
+
+    @Override
+    public boolean equals(final Object otherVersionId) {
+        if (Objects.isNull(otherVersionId) || !(otherVersionId instanceof VersionId)) {
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: I don't see that. Which version of Assert do you have in mind? ;-)

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            return false;
+        }
+        VersionId other = (VersionId) otherVersionId;
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: done

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: done

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    /**
+     * Check if {@code otherVersionId} is a match with this version-id considering
+     * {@link #hasPrefixMatchModifier()} and {@link #hasGreaterThanOrEqualMatchModifier()}.
+     *
+     * @param otherVersionId a version-id
+     * @return {@code true} if this version-id matches {@code otherVersionId}, {@code false} otherwise.
+     */
+    public boolean matches(final VersionId otherVersionId) {
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: done

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: oh, sorry. Missed the second part of the if statement...

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        VersionId other = (VersionId) otherVersionId;
+
+        if (compoundVersionId != null) {
+            return this.compoundVersionId.equals(other.compoundVersionId == null? other : other.compoundVersionId);
+        }
+
+        final String[] tuple1 = this.asNormalizedTuple(other.asTuple().length);
+        final String[] tuple2 = other.asNormalizedTuple(this.asTuple().length);
+
+        for (int i = 0; i < tuple1.length; i++) {
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: I don't think this makes the code more readable here, as I need the index here.

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        }
+
+        final String[] tuple1 = this.asTuple();
+        final String[] tuple2 = otherVersionId.asNormalizedTuple(tuple1.length);
+
+        for (int i = 0; i < tuple1.length; i++) {
+            final Object tuple1Element = prepareForNormalizedComparision(tuple1[i]);
+            final Object tuple2Element = prepareForNormalizedComparision(tuple2[i]);
+
+            if (tuple1Element.equals(tuple2Element)) {
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: done

File: core/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/JreVersion.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            LOG.info("good - your JRE - {} - match requested JRE - {}", getJreVersion(), v);
+        }
+    }
+
+    public boolean matchesJreVersion() {
+        return matches(getJreVersion());
+    }
+
+    private String getJreVersion() {
+        return System.getProperty("java.version");
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply core/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/JreVersion.java]: done

File: common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+     *
+     * @param tupleValue the raw tuple value
+     * @return the tuple value suited for normalized comparision
+     */
+    private static Object prepareForNormalizedComparision(final String tupleValue) {
+        if (tupleValue.length() > 0 && tupleValue.charAt(0) != '-') {
+            try {
+                return Integer.valueOf(tupleValue);
+            }
+            catch (NumberFormatException ex) { /* not numeric, as not parsable as Java int */ }
```
[the remark this answers is not in the document]
Comment (AndreasEhret) [reply common/src/main/java/net/adoptopenjdk/icedteaweb/jnlp/version/VersionId.java]: actually it is handled by not handling here. But I made this more understandable and fail-safe at the cost of the extra return in the catch block and better comment.

**Pull request merged**